jax_distribution/evaluator.py

```diff
@@ -124,35 +125,185 @@ def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOut
+        return eval_metrics
+
+    def evaluator_fn(trained_params: FrozenDict, rng: chex.PRNGKey) -> ExperimentOutput:
+        """Evaluator function."""
+
+        # Initialise environment states and timesteps.
+        n_devices = len(jax.devices())
+
+        eval_batch = config["num_eval_episodes"] // n_devices * eval_multiplier
+
+        rng, *env_rngs = jax.random.split(rng, eval_batch + 1)
+        env_states, timesteps = jax.vmap(env.reset)(
+            jnp.stack(env_rngs),
+        )
+        # Split rngs for each core.
+        rng, *step_rngs = jax.random.split(rng, eval_batch + 1)
+        # Add dimension to pmap over.
+        reshape_step_rngs = lambda x: x.reshape(eval_batch, -1)
+        step_rngs = reshape_step_rngs(jnp.stack(step_rngs))
+
+        # Initialise hidden state.
+        init_hstate = ScannedRNN.initialize_carry(eval_batch, 128)
+        init_hstate = jnp.expand_dims(init_hstate, axis=1)
+        init_hstate = jnp.expand_dims(init_hstate, axis=2)
+        init_hstate = jnp.tile(init_hstate, (1, config["num_agents"], 1))
+
+        # Initialise dones.
+        dones = jnp.zeros(
+            (
+                eval_batch,
+                config["num_agents"],
+            ),
+            dtype=bool,
+        )
+
+        eval_state = RNNEvalState(step_rngs, env_states, timesteps, dones, init_hstate)
+
+        eval_metrics = jax.vmap(
+            eval_one_episode, in_axes=(None, 0), axis_name="eval_batch"
+        )(trained_params, eval_state)
+
+        return ExperimentOutput(
+            episodes_info=eval_metrics,
+        )
+
+    return evaluator_fn
+
+
 def evaluator_setup(
     eval_env: Environment,
     rng_e: chex.PRNGKey,
     network: Any,
     params: FrozenDict,
     config: Dict,
+    rnn_net: bool = False,
+    ScannedRNN: nn.Module = None,
 ) -> Tuple[callable, callable, Tuple]:
     """Initialise evaluator_fn, network, optimiser, environment and states."""
     # Get available TPU cores.
     n_devices = len(jax.devices())
 
-    # Vmap it over number of agents.
-    vmapped_eval_network_apply_fn = jax.vmap(
-        network.apply,
-        in_axes=(None, 0),
-    )
-
-    evaluator = get_evaluator_fn(
-        eval_env,
-        vmapped_eval_network_apply_fn,
-        config,
-    )
-
-    absolute_metric_evaluator = get_evaluator_fn(
-        eval_env,
-        vmapped_eval_network_apply_fn,
-        config,
-        10,
-    )
+    # Vmap it over number of agents and create evaluator_fn.
+    if rnn_net:
+        vmapped_eval_network_apply_fn = jax.vmap(
+            network.apply, in_axes=(None, 1, 2), out_axes=(1, 2, 2)
+        )
+        evaluator = get_rnn_evaluator_fn(
+            eval_env,
+            vmapped_eval_network_apply_fn,
+            config,
+            ScannedRNN,
+        )
+        absolute_metric_evaluator = get_rnn_evaluator_fn(
+            eval_env,
+            vmapped_eval_network_apply_fn,
+            config,
+            ScannedRNN,
+            10,
```

Comment:
The absolute metric is suggested to be 10. So I don't even think we need it in the config. 
Maybe just a comment to say that this is 10 because of the default absolute metric requirements.